### ISSUE TITLE
Vna.py: Creating segmented sweeps with over 5000 segments may cause timeout

### ISSUE THREAD

**Issue by Terrabits** [Terrabits/rohdeschwarz#3]
It seems that the ZVA and ZNB firmware has a hard time handling segmented (frequency) sweeps where the number of segments is well beyond the normal range. At approximately 5000 segments, the firmware takes longer and longer to process each new added segment. Because of this sluggishness, a timeout may occur.

This effects the `VnaChannel::frequencies_Hz=` method and manual use of the `SENS<Ch>:SEGM<i>:ADD` SCPI command.

Unfortunately, this is not something that I can fix in the package itself. I document it here as a bug to prevent others from wasting time on this. I have also reported this bug to the development team. I will update if/when there is a plan to fix it.

The only work-around is to increase the timeout time and be patient.
`GenericInstrument::timeout_ms=<long_time>`

**Comment by onealj** [comment]
Not sure if the ZVA and ZNB offer it, but on the FSW, I query the sweep time based on the current settings before executing a sweep, increase the timeout if necessary, execute the sweep, the restore the original timeout.

I can justify temporarily changing the timeout without the user explicitly increasing it because usually you want a timeout error to occur when you expect a response and don't receive one. During long sweeps, I don't expect a response until the sweep is complete, and I don't expect the user to micromanage the timeout when they execute every sweep to check the timeout. That's why I buried the timeout increaser in the instrument driver.

Using my internal library for a executing a sweep on a Rohde and Schwarz FSW. All delays are given in seconds, not milliseconds.

    # some methods omitted for brevity
    class RohdeSchwarzFSW(Vxi11Instrument):
        def sweep(self):
            self.send('INIT')
            swcount = max(self.swcount, 1)
            duration = max(self.swduration, 2) * count
            with self.increase_timeout(2*duration + 2.0):
                self.query('*OPC?')

        def swduration(self):
            # Estimated duration of a single sweep, in seconds
            # includes measurement (SENSE:SWEEP:TIME) plus any processing time
            return float(self.query('SENSe:SWEep:DURation?'))

        @contextmanager
        def increase_timeout(self, new_timeout):
            #__enter__:
            try:
                # set timeout to the larger of the current timeout and the requested timeout
                old_timeout = self.timeout
                if new_timeout > old_timeout:
                    self.set_timeout(new_timeout)
                yield

            #__exit__:
            finally:
                # restore original timeout
                if new_timeout > old_timeout:
                    self.set_timeout(old_timeout)

    sa = RohdeSchwarzFSW(ip_address)
    sa.open()
    sa.preset()
    sa.sweep_continuous = False
    sa.startfreq = 0 #Hz
    sa.stopfreq = 1e9 #Hz
    sa.swpoints = 100001
    sa.rbw = 1 #Hz
    sa.swcount = 100
    sa.trace_mode[1] = 'AVER'
    # trigger a long measurement
    sa.sweep()
    # fetch the result from trace 1
    x, y = sa.trace[1]
    plot(x, y)

**Comment by Terrabits** [comment]
Hi @onealj,

Thanks for the suggestion. I actually do that in my VNA driver when handling sweeps. I hadn't considered using context managers. I think my code would be more legible with this mechanism rather than manually changing timeout there-and-back.

tl;dr:
**R&S VNAs become sluggish with too many segments. They become less and less usable in ways beyond timeout issues.**

The problem with the R&S VNAs is that they "lock up", seemingly starting at around 5k segments. This effects the manual interface as well. I don't think the designers envisioned a need for that many segments; the firmware definitely does not like it.

Increasing the timeout would accommodate this to some degree, but you'd have to increase the timeout for all subsequent (even unrelated) operations to the point where the user would probably get frustrated and give up. So I decided to leave the library the way it is and hope that the firmware is improved at some point in the future.

The best work-around is to avoid creating too many segments if possible. For example, the VnaChannel.frequency_Hz= setter takes a list of frequencies and creates as many segments, each with that frequency as the sole point. While this is very flexible and works for all frequency axis in theory, because of the aforementioned bug it would be best to use linearly- and logarithmically defined sweeps when possible.

For example, `vna.channel(1).frequency_Hz = numpy.linspace(1e9, 2e9, 201)` would instead become:

```python
ch1        = vna.channel(1)
ch1.start  = 1e9
ch1.stop   = 2e9
ch1.points = 201
```

If you do need a segmented sweep, the same concept applies to each segment: try to use as few linear or log-spaced segments as possible, each with as many points as possible.
